Manage metadata registry

samuelcambien opened the pull request:
This PR covers the CRUD operations on both the MetadataField and MetadataSchema endpoints for the REST api:

* The createAndReturn (create), put (update) and delete (delete) methods have been implemented in the MetadataFieldRestRepository and the MetadataSchemaRestRepository.
* Additional tests have been added to the MetadataFieldRestRepositoryIT and MetadataSchemaRestRepositoryIT to keep the functionality properly tested.

Jira link: https://jira.duraspace.org/browse/DS-4109
RestContract PR link: https://github.com/DSpace/Rest7Contract/pull/38

Commit:
Added support for the CRUD operations on the MetadataField and MetadataSchema REST endpoints

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -407,4 +408,20 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
         throw new RepositoryMethodNotImplementedException("No implementation found; Method not allowed!", "");
     }
 
+    public T put(HttpServletRequest request, String apiCategory, String model, ID uuid, JsonNode jsonNode) {
```
Javadocs please (all other methods in the class include javadocs).

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -981,4 +981,38 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+
+
+    @RequestMapping(method = RequestMethod.PUT, value = REGEX_REQUESTMAPPING_IDENTIFIER_AS_DIGIT)
+    public DSpaceResource<RestAddressableModel> put(HttpServletRequest request,
```
Javadocs please

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -981,4 +981,38 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+    }
+
+    @RequestMapping(method = RequestMethod.PUT, value = REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID)
+    public DSpaceResource<RestAddressableModel> put(HttpServletRequest request,
```
Javadocs please

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -981,4 +981,38 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+        return putOneInternal(request, apiCategory, model, uuid, jsonNode);
+    }
+
+    private <ID extends Serializable> DSpaceResource<RestAddressableModel> putOneInternal(HttpServletRequest request,
```
Javadocs please

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java:
```diff
@@ -98,4 +113,116 @@ public MetadataFieldRest findOne(Context context, Integer id) {
+
+        // parse request body
+
+        MetadataFieldRest metadataFieldRest;
```
Thanks for the inline comments!  Just a minor thing, but the empty line after each of the inline comments is not necessary (I noticed that occurs throughout this method).

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:
```diff
@@ -73,4 +87,102 @@ public MetadataSchemaRest findOne(Context context, Integer id) {
+
+        // parse request body
+
+        MetadataSchemaRest metadataSchemaRest;
```
Again, empty lines after comments in this method can be removed.

tdonohue reviewed (commented):
Did an initial, quick code review. A few areas of missing Javadocs found.

Also, this PR is currently failing compilation in Travis. The issues look to be minor though.

Commit:
Added support for the CRUD operations on the MetadataField and MetadataSchema REST endpoints - feedback

samuelcambien commented:
Hi Tim, thanks for your feedback, I pushed a new commit addressing your remarks.
About the empty lines after the inline comments, I added them because the comments relate to the whole block of code following them, not just to the single following line, so I thought that layout would make this more clear. But anyway I removed them now :)
Normally Travis should succeed now too..

benbosman commented:
This PR will also be updated based on the feedback in https://github.com/DSpace/Rest7Contract/pull/38

KevinVdV commented:
I merged the master branch into this PR to solve the issues with Travis. The changes mentioned in DSpace/Rest7Contract#38 are already present here. This PR is now ready for review.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -407,4 +408,44 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
+            thisRepository.put(context, request, apiCategory, model, uuid, jsonNode);
+            context.commit();
+        } catch (SQLException | AuthorizeException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
`e.getMessage()` can be null, which would make this RuntimeException uninformative.  Please replace `e.getMessage()` with a more useful error message.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java:
```diff
@@ -98,4 +113,112 @@ public MetadataFieldRest findOne(Context context, Integer id) {
+                    MetadataFieldRest.class
+            );
+        } catch (IOException excIO) {
+            throw new PatchBadRequestException("error parsing the body ..." + excIO.getMessage());
```
This is swallowing the underlying stacktrace, and just returning `getMessage()` (which can be null). This should probably instead be: `throw new PatchBadRequestException("error parsing the body", excIO);`

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java:
```diff
@@ -98,4 +113,112 @@ public MetadataFieldRest findOne(Context context, Integer id) {
+
+            metadataFieldService.delete(context, metadataField);
+        } catch (SQLException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
Again, `getMessage()` can return null.  Please replace with a more informative error message.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:
```diff
@@ -73,4 +87,98 @@ public MetadataSchemaRest findOne(Context context, Integer id) {
+                    MetadataSchemaRest.class
+            );
+        } catch (IOException excIO) {
+            throw new PatchBadRequestException("error parsing the body ..." + excIO.getMessage());
```
Again, swallowing/hiding the stacktrace.  This should be:
`throw new PatchBadRequestException("error parsing the body", excIO);`

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:
```diff
@@ -73,4 +87,98 @@ public MetadataSchemaRest findOne(Context context, Integer id) {
+
+            metadataSchemaService.delete(context, metadataSchema);
+        } catch (SQLException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
Again, `getMessage()` should be replaced with a more informative error

tdonohue reviewed (commented):
Gave it another code review today, and noticed some of the exception handling needs cleanup work.

Commit:
Added support for the CRUD operations on the MetadataField and MetadataSchema REST endpoints - feedback

samuelcambien commented:
Hi Tim, thanks for your feedback. I addressed the issues in my last commit.

tdonohue committed:
Minor improvements to English in error messages.

tdonohue committed:
More minor corrections to English

tdonohue committed:
English grammar corrections

tdonohue reviewed (approved):
👍 I gave this another review today, and committed some minor corrections to English grammar in error messages.  It looks good to me.  Thanks again @samuelcambien !

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+    }
+
+    @Test
+    public void createUnauthauthorized() throws Exception {
```
please fix the typo in the method name

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+            deleteMetadataFieldIfExists();
+        }
+    }
+
```
add a test for the forbidden scenario (loggedin user without  admin rights)

tdonohue replied:
This test was added.  See `createUnauthorizedEPersonNoAdminRights()` in the code.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+            deleteMetadataFieldIfExists();
+        }
+    }
+
```
add a test for the forbidden scenario (loggedin user without  admin rights)

tdonohue replied:
This test was added. See `deleteUnauthorizedEPersonNoAdminRights()` in the code

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+        }
+    }
+
+    private MetadataField createMetadataField() throws AuthorizeException, SQLException, NonUniqueMetadataException {
```
we should have a Builder for that

tdonohue replied:
This has been updated to use existing `MetadataFieldBuilder` class.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+            context.turnOffAuthorisationSystem();
+            metadataFieldService.delete(context, metadataField);
+            context.commit();
+        }
```
to be safe you should restore the authorisationSystem

tdonohue replied:
This code seems to be refactored and no longer exists.  I believe it was replaced with the `MetadataFieldBuilder` class.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,255 @@
 import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
 
 public class MetadataSchemaRestRepositoryIT extends AbstractControllerIntegrationTest {
```
Please provide basic javadoc for all the public methods summarizing the test condition and the expected result

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -7,23 +7,62 @@
 import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
 
 public class MetadatafieldRestRepositoryIT extends AbstractControllerIntegrationTest {
```
Please provide basic javadoc for all the public methods summarizing the test condition and the expected result

abollini reviewed (dismissed):
thanks for the PR. The code looks aligned with the agreed rest contract. I have reported some required cleanups inline in the code

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -97,7 +98,7 @@
     /**
      * Return a specific REST object
-     * 
+     *
```
I'd consider this very low priority right now.  Not sure if it's easy to check for in code style (and even if it is, it might take significant effort to align all comments).  So, I'm resolving this comment.

Raf-atmire committed:
Applied feedback and fixed test cases

Raf-atmire commented:
Hi @abollini thank you for your feedback. I've looked into your feedback and fixed all the issues in my latest commit. However I believe that fully documenting test classes with javadocs, when we're talking about really simple tests that don't cover a lot of ground, bloats the entire class and adds really little value as the information should be clear from just the test name. For example, this hasn't been done for EPersonRestRepositoryIT, GroupRestRepositoryIT and so on.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -985,4 +985,86 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+    /**
+     * Execute a PUT request for an entity with id of type Integer;
+     *
+     * curl -X PUT http://<dspace.url>/dspace-spring-rest/api/{apiCategory}/{model}
```
I double checked and these were updated. Marking this request as "resolved"

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -985,4 +985,86 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+     * Example:
+     * <pre>
+     * {@code
+     *      curl -X PUT http://<dspace.url>/dspace-spring-rest/api/metadatafield
```
This is now resolved as well.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -985,4 +985,86 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+    /**
+     * Execute a PUT request for an entity with id of type UUID;
+     *
+     * curl -X PUT http://<dspace.url>/dspace-spring-rest/api/{apiCategory}/{model}
```
Also resolved

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -407,4 +408,46 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
     }
 
+    /**
+     * Method to support updating a DSpace instance.
```
This text was updated.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -407,4 +408,46 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
+    }
+
+    /**
+     * Implement this method in the subclass to support updating a DSpace instance.
```
Text was updated

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -124,4 +163,190 @@ public void findByNullSchema() throws Exception {
+                    .andExpect(status().isCreated())
+                    .andDo(result -> idRef.set(read(result.getResponse().getContentAsString(), "$.id")));
+
+            MetadataField metadataField = metadataFieldService.find(context, idRef.get());
```
This test has been updated to perform a `GET` request to ensure object was created. Marking this comment as resolved

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MetadataSchemaRest.java:
```diff
@@ -7,13 +7,15 @@
  *
  * @author Andrea Bollini (andrea.bollini at 4science.it)
  */
+@JsonIgnoreProperties(ignoreUnknown = true)
```
This annotation was removed from code.  Marking this comment as resolved.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
         context.turnOffAuthorisationSystem();
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, "ATest", "ANamespace")
+                .build();
+
```
We should be turning back *on* the AuthorizationSystem here, shouldn't we?  We only need it off temporarily in order to create test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+    @Test
+    public void createUnauthauthorizedTest()
+            throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.  We aren't creating test data in this method.

Also, this method name has a misspelling. It should be `createUnauthorizedTest()`

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+        context.turnOffAuthorisationSystem();
+
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, TEST_NAME, TEST_NAMESPACE)
+                                                             .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+        context.turnOffAuthorisationSystem();
+
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, "A name", "A namespace")
+                                                             .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+        context.turnOffAuthorisationSystem();
+
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, TEST_NAME, TEST_NAMESPACE)
+                                                             .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+        context.turnOffAuthorisationSystem();
+
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, TEST_NAME, TEST_NAMESPACE)
+                                                             .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
 
+    @Test
+    public void createSuccess() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadataSchemaRestRepositoryIT.java:
```diff
@@ -7,50 +7,218 @@
+        context.turnOffAuthorisationSystem();
+
+        MetadataSchema metadataSchema = MetadataSchemaBuilder.createMetadataSchema(context, "ATest", "A namespace")
+                                                             .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -65,13 +100,15 @@ public void findOne() throws Exception {
 
     @Test
     public void searchMethodsExist() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -108,6 +145,7 @@ public void findBySchema() throws Exception {
 
     @Test
     public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
 
     @Test
     public void findByNullSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+
+    @Test
+    public void createUnauthorized() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+
+    @Test
+    public void createUnauthorizedEPersonNoAdminRights() throws Exception {
+        context.turnOffAuthorisationSystem();
```
I believe the Auth system should remain on in this method.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/MetadatafieldRestRepositoryIT.java:
```diff
@@ -119,9 +157,201 @@ public void findByUndefinedSchema() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        MetadataField metadataField = MetadataFieldBuilder.createMetadataField(context, ELEMENT, QUALIFIER, SCOPE_NOTE)
+                                                          .build();
```
We should turn the auth system back on after creating the test data.

tdonohue reviewed (changes requested):
@samuelcambien : I gave this a review today.  The actual code is looking good overall, but the Integration Tests seem to be turning off the DSpace Authorization System (and leaving it off) in almost every method.  We should *only* be turning off authorization for temporary purposes (like creating test data), otherwise we run the risk of having code that passes Integration Tests (cause authorization isn't being checked) but fails in manual testing.

tdonohue commented:
@Raf-atmire & @abollini:  I've done a review of existing ITs for the REST API, and while we've been *inconsistent* about requiring them, many of our IT classes *do have* comments at the **class level** to describe which endpoint the tests work against, etc.   See for example: OpenSearchControllerIT, AuthenticationRestControllerIT, BitstreamContentRestControllerIT, IdentifierRestControllerIT, RestResourceControllerIT, RootRestResourceControllerIT, SubmissionDefinitionsControllerIT, etc etc.  That said, I *have* noticed a trend that most `*ControllerIT` classes have these comments, while most `*RestRepositoryIT` classes do not.

However, since the IT classes in this PR are **not** brand new classes, I don't think we should require this PR to correct the mistakes of past PRs.  That said, honestly, it's 3-4 lines of comments that are being requested, so if you have the chance to add them, it'd be appreciated.

In next week's DSpace 7 meeting will will be discussing/finalizing requirements for comments in ITs.  So, please be aware that this could affect the code review of all future PRs.

paulo-graca commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -987,20 +987,48 @@ private void addLocationHeader(HttpHeaders headers, PersistentEntityResourceAsse
+                                                    @PathVariable String apiCategory, @PathVariable String model,
+                                                    @PathVariable Integer id,
+                                                    @RequestBody JsonNode jsonNode) {
+        return putOneInternal(request, apiCategory, model, id, jsonNode);
```
This isn't a specific issue about this PR. I've tried to do a metadatafield or metadataschema PUT requests when the user doesn't have the correct permissions and I get an 404 error. I was expecting to get a 403 Forbidden instead.

Raf-atmire replied:
I can't seem to reproduce the described issue. I've written extra tests to perform the exact behaviour that was described and they succeed: https://github.com/DSpace/DSpace/pull/2291/commits/d450b59899119cd70e32c3c2f788c22c19d0d02a#diff-9e02166d9f889684bb5eba075a991012R373
https://github.com/DSpace/DSpace/pull/2291/commits/d450b59899119cd70e32c3c2f788c22c19d0d02a#diff-d2e1fa933aba2c21b6d604cdbece20e5R239

paulo-graca replied:
You can test with the Atmire's database - https://www.dropbox.com/s/ovqp394y3vofnwa/entities7-test-db.sql.gz?dl=1
And with a user that doesn't have any record on epersongroup2eperson, like atmirenv+write5a@g****.com.

Trying this PUT http://localhost/spring-rest/api/core/metadataschemas/1
`{"id":1,"prefix":"dc","namespace":"http://dublincore.org/documents/dcmi-terms/"}`

I've got an 404 error, please check out my apache log:

`SERVER - - [11/Feb/2019:15:14:37 +0000] "PUT /spring-rest/api/core/metadataschemas/1 HTTP/1.1" 404 949 "http://localhost/admin/registries/metadata/eperson?pageId=registry-metadatafields-pagination&page=1&pageSize=25&sortDirection=ASC&sortField=id" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"`

But, like I said, I think this is an Inherited problem, not specific to this PR.

tdonohue replied:
@paulo-graca and @Raf-atmire : I've logged this as a bug in this JIRA ticket: https://jira.duraspace.org/browse/DS-4168

As noted in today's meeting, it seems like this bug is unrelated to this specific PR.  So, I'd encourage someone to claim that JIRA ticket and look into it.

Raf-atmire committed:
Applied feedback and added tests

Raf-atmire commented:
@tdonohue Thank you for the review. I've addressed the issues with the context authorisation system by turning it back on after creating test data or removing the line completely. I do believe that there is a misconception as turning off the authorisation system in these test classes won't affect the REST call's behaviour since otherwise none of the 'forbidden' or 'accessdenied' tests would have succeeded, therefor this should never affect the actual REST call that we're going to test.
I've also added a small JavaDoc section above the two IT classes as requested.
Additionally I've added two tests to provide a test case for the issue that @paulo-graca had run into.

tdonohue reviewed (approved):
👍 I've re-reviewed this today.  All of my suggested changes have been made, as have all of @abollini's suggested changes.  This looks good to me.  Thanks for the updates @Raf-atmire

tdonohue commented:
We discussed this PR in today's DSpace 7 Meeting and decided it was ready to merge.  @paulo-graca's notes about a 404 error response seem unrelated to this PR, and I've logged a new JIRA ticket for them at https://jira.duraspace.org/browse/DS-4168
